File: algorithms/javaalgo/src/test/java/programmers/level2/lifeboat/LifeBoatTest.java
package programmers.level2.lifeboat;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

import java.util.Arrays;
import org.junit.jupiter.api.Test;

public class LifeBoatTest {

    @Test
    void test() {
        int[] people1 = {70, 50, 80, 50};
        int[] people2 = {70, 80, 50};
        assertAll(
            () -> assertThat(solution(people1, 100)).isEqualTo(3),
            () -> assertThat(solution(people2, 100)).isEqualTo(3)
        );
    }

    public int solution(int[] people, int limit) {
        int answer = people.length;
        Arrays.sort(people);
        int right = people.length - 1;

        for (int left = 0; left < people.length; left++) {
            while (left < right) {
                if (people[left] + people[right] <= limit) {
                    answer--;
                    right--;
                    break;
                } else {
                    right--;
                }
            }
        }
        return answer;
    }
}

File: algorithms/javascriptalgo/programmers/level2/구명보트/lifeboat.test.js
test("test", () => {
  let people1 = [70, 50, 80, 50];
  let people2 = [70, 80, 50];

  expect(solution(people1, 100)).toBe(3);
  expect(solution(people2, 100)).toBe(3);
});

function solution(people, limit) {
  let answer = 0;
  people.sort((a, b) => b - a);
  let left = 0;
  let right = people.length - 1;

  while (left < right) {
    if (people[left] + people[right] > limit) {
      left++;
    } else {
      left++;
      right--;
    }
    answer++;
  }

  if (left === right) {
    answer++;
  }

  return answer;
}
